repo/fsrepo/migrations/migrations.go

```diff
@@ -1,282 +1,208 @@
+	} else {
+		log.Println("  => Running:", binPath, pathArg, "-verbose=true")
+		cmd = exec.CommandContext(ctx, binPath, pathArg, "-verbose=true")
 	}
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	return cmd.Run()
+}
 
-	var archive string
-	switch runtime.GOOS {
-	case "windows":
-		archive = "zip"
-		binnom += ".exe"
-	default:
-		archive = "tar.gz"
-	}
+// fetchMigrations downloads the requested migrations, and returns a slice with
+// the paths of each binary, in the same order specified by needed.
+func fetchMigrations(ctx context.Context, needed []string, destDir string) ([]string, error) {
 	osv, err := osWithVariant()
 	if err != nil {
-		return err
+		return nil, err
 	}
-
 	if osv == "linux-musl" {
-		return fmt.Errorf("linux-musl not supported, you must build the binary from source for your platform")
-	}
-
-	finame := fmt.Sprintf("%s_%s_%s-%s.%s", distname, vers, osv, runtime.GOARCH, archive)
-	distpath := fmt.Sprintf("%s/%s/%s/%s", root, distname, vers, finame)
-
-	data, err := httpFetch(distpath)
-	if err != nil {
-		return err
-	}
-
-	arcpath := filepath.Join(dir, finame)
-	fi, err := os.Create(arcpath)
-	if err != nil {
-		return err
-	}
-
-	_, err = io.Copy(fi, data)
-	if err != nil {
-		return err
-	}
-	fi.Close()
-
-	return unpackArchive(distname, binnom, arcpath, out, archive)
-}
-
-// osWithVariant returns the OS name with optional variant.
-// Currently returns either runtime.GOOS, or "linux-musl".
-func osWithVariant() (string, error) {
-	if runtime.GOOS != "linux" {
-		return runtime.GOOS, nil
-	}
-
-	// ldd outputs the system's kind of libc.
-	// - on standard ubuntu: ldd (Ubuntu GLIBC 2.23-0ubuntu5) 2.23
-	// - on alpine: musl libc (x86_64)
-	//
-	// we use the combined stdout+stderr,
-	// because ldd --version prints differently on different OSes.
-	// - on standard ubuntu: stdout
-	// - on alpine: stderr (it probably doesn't know the --version flag)
-	//
-	// we suppress non-zero exit codes (see last point about alpine).
-	out, err := exec.Command("sh", "-c", "ldd --version || true").CombinedOutput()
-	if err != nil {
-		return "", err
+		return nil, fmt.Errorf("linux-musl not supported, you must build the binary from source for your platform")
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(len(needed))
+	bins := make([]string, len(needed))
+	// Download and unpack all requested migrations concurrently.
+	for i, name := range needed {
+		log.Printf("Downloading migration: %s...", name)
+		go func(i int, name string) {
+			defer wg.Done()
+			distDir := path.Join(distMigsRoot, name)
+			ver, err := LatestDistVersion(ctx, distDir)
+			if err != nil {
+				log.Printf("could not get latest version of migration %s: %s", name, err)
+				return
+			}
+			loc, err := FetchBinary(ctx, distDir, ver, name, name, destDir)
+			if err != nil {
+				log.Printf("could not download %s: %s", name, err)
+				return
+			}
+			log.Printf("Downloaded and unpacked migration: %s (%s)", loc, ver)
+			bins[i] = loc
```

Comment:
Yes, this is completely safe because:
- Slice is never resized
- Each goroutine has a separate element to write to; same as if each wrote to separate closed-over variable
- Slice is only read after passing `wg.Wait()` synchronization point

This is a safe and efficient implementation.  It is useful when there is a separate place for each goroutine to write its output, and there is nothing for the reader to do until all goroutines finish.  

The tl;dr version:
If multiple goroutines access a variable concurrently, and at least one of the accesses is a write, then synchronization is required.  Here the rule is not violated, because the goroutines do not write the slice value (slice header), they only read it when indexing the slice. Each goroutine only modifies a single, different, designated slice element. Since each slice element has its own address (own memory space), they are like distinct variables. [Spec: Variables](https://golang.org/ref/spec#Variables) says,
>Structured variables of [array](https://golang.org/ref/spec#Array_types), [slice](https://golang.org/ref/spec#Slice_types), and [struct](https://golang.org/ref/spec#Struct_types) types have elements and fields that may be [addressed](https://golang.org/ref/spec#Address_operators) individually. Each such element acts like a variable.

Follwoing the rule, results cannot be read without synchronization.  The `waitgroup` provides sufficient synchronization because it ensures all goroutines complete and cannot modify elements after `wg.Wait()` returns, and ensures the modification are visible to the goroutine that called `wg.Wait()`